### ISSUE TITLE
Question: Which rules CLR JIT uses to place VZEROUPPER ?

### ISSUE THREAD

**Issue by voinokin** [dotnet/coreclr#18400]
Which rules CLR JIT uses to place VZEROUPPER, and how do I opt-out ?

The live example is - one flavor of distribution sorting algorithm uses insertion sort subroutine for basic case sorting of small buckets, and it gets called quite often esp. for of 2- and 3-record cases.  The insertion sort subroutine itself is highly optimized - it uses SIMD registers (XMMs in my case), HW intrinsics and just basic RSP-based stack frame for couple values. Still VZEROUPPER is generated in the beginning and in the end of subroutine, which adds some perfmormance hit. There is natural desire to avoid such penalty knowing that caller will always be other CLR JITted code, and all nested method invocations are inlined (checked this manually).

**Comment by jkotas** [comment]
The rules are in a comment in front of `CodeGen::genVzeroupperIfNeeded`: https://github.com/dotnet/coreclr/blob/master/src/jit/codegencommon.cpp#L10235

**Comment by voinokin** [comment]
Thanks! According to the link:

> Issue VZEROUPPER in Prolog if the method contains 128-bit or 256-bit AVX code, to avoid legacy SSE to AVX transition penalty, which could happen when native code contains legacy SSE code calling into JIT AVX code (e.g. reverse pinvoke). Issue VZEROUPPER in Epilog if the method contains 256-bit AVX code, to avoid AVX to legacy SSE transition penalty.

To be precise the comment should tell not about mixing SSE and AVX code, but rather about mixing non-VEX and VEX-encoded code. VEX-encoded instructions operating on XMM as lower part of Y/ZMM register **DO IMPLICIT ZEROEING** of upper part and eliminate false data dependencies.
This link states example difference between handling of "Legacy SSE instruction" and "VEX.128 encoded instruction": http://www.felixcloutier.com/x86/PADDB:PADDW:PADDD:PADDQ.html
See also discussion on `VZEROUPPER` here: https://software.intel.com/en-us/forums/intel-isa-extensions/topic/704023

I wonder whether there are any cases other than reverse pinvoke, otherwise inserting `VZEROUPPER` everywhere looks more like overkill since most of the time the JITted code will be called by other JITted code, and will return there too. It is of course still to confirm whether JIT will **ALWAYS** generate just VEX encoded instructions if VEX is available, or can intermix non-VEX and VEX for some reason.
The reverse pinvoke itself along with top-level code like `Main()` could be handled separately, eg. when unmanaged->managed trampoline is constructed.

**Comment by voinokin** [comment]
Related: #18144

**Comment by benaadams** [comment]
> The history is this: The extension of vector registers from 128 to 256 bits caused a problem when legacy Windows device drivers saved only the lower 128 bits of the new 256-bit registers. This problem was solved in a rather complex way. The Sandy Bridge processor could switch between a VEX state with full 256-bit registers and a non-VEX state where all the 256-bit registers were split into two 128-bit parts. The switch between these states had a cost of 70 clock cycles. The instruction VZEROUPPER was used for avoiding the cost of this state transition by clearing the upper half of all the registers. Alternatively, one could use VZEROALL to clear the whole registers. The code has to use VZEROUPPER after any code that uses 256-bit registers if there is any chance that the subsequent code contains non-VEX vector instructions. The recommendation from Intel was to use VZEROUPPER in AVX code before any call or return to an ABI-compliant function with unknown VEX status.


> * The following processors have expensive state transitions and cheap VZEROUPPER and VZEROALL: Sandy Bridge, Ivy Bridge, Haswell, and Broadwell. VZEROUPPER is needed for performance reasons on these processors.
> * There is no expensive state transition on the later Intel processors: Skylake and Knights Landing.
> * There is no expensive state transition on AMD processors.
> * VZEROUPPER and VZEROALL are expensive on Knights Landing. I have measured 36 clock cycles for both instructions in 64-bit mode (30 clock in 32-bit mode).

... moving on to latest being

> Now that Skylake processors with AVX512 are available, we are able to verify how it works. My tests show the following:
>
> The processor will switch to the dirty state when YMM0-YMM15 or ZMM0-ZMM15 are touched, but not when ZMM16-ZMM31 are touched. It switches back to the clean state when VZEROUPPER or VZEROALL is executed. You can avoid state transitions and penalties by using ZMM16-ZMM31 only. Of course this doesn't work if vector registers are used for function parameters or function returns when the calling convention dictates that the lower vector registers must be used.
>
> SSE instructions that write to an XMM register have a false dependence on any previous write to the same register when the processor is in the dirty state. This affects mostly move instructions, while most other SSE instructions have the destination register as input anyway. The performance cost is small in most cases. I cannot confirm the 2x to 6x slowdown that Travis claims.
>
> The conclusion is: Use VZEROUPPER when leaving VEX code, or use ZMM16-ZMM31 only. You may need to do differently on Knights Landing where VZEROUPPER is expensive.

**Comment by tannergooding** [comment]
CC. @CarolEidt.

Given that this does only impact interop code that executes SIMD, maybe we could allow the user to explicitly opt-out of having these emitted by default (they could then emit it themselves via an explicit intrinsic)?

Another idea would be to only emit this before or after interop call-sites.

**Comment by gfoidl** [comment]
>  interop code that executes SIMD

When looking for a solution to this, it has to kept in mind that `A` might interop with `B`, and `A` doesn't know something about the internal working of `B` -- i.e. whether `B` uses SIMD or not.

Therefore with 
> emit this before or after interop call-sites.

it may be on the safe side.

**Comment by mikedn** [comment]
> I wonder whether there are any cases other than reverse pinvoke, otherwise inserting VZEROUPPER everywhere looks more like overkill since most of the time the JITted code will be called by other JITted code, and will return there too

Yes, there are other cases - crossgened code. Crossgen only uses non-VEX SSE so such transitions are possible without involving interop.

>  It is of course still to confirm whether JIT will ALWAYS generate just VEX encoded instructions if VEX is available, or can intermix non-VEX and VEX for some reason.

Yes, at runtime the JIT always generate VEX/non-VEX code. At least in the today world where there's a single JIT. If an alternative JIT is ever used then things could be different. Though it wouldn't make much sense to have 2 JIT that behave differently in this regard.

**Comment by voinokin** [comment]
> > It is of course still to confirm whether JIT will ALWAYS generate just VEX encoded instructions if VEX is available, or can intermix non-VEX and VEX for some reason.

> Yes, at runtime the JIT always generate VEX/non-VEX code

Do you mean that JIT can produce both VEX and non-VEX code mixed on the same machine within the same process instance ?

**Comment by CarolEidt** [comment]
@mikedn has it correct that we need to be concerned with crossgened code as well as interop.

One option to consider, in addition to, or instead of an "opt out" approach would be to emit only the vzeroupper instructions before & after interop call-sites when `COMPlus_ZapDisable` is set. That forces all methods to be JITed which would avoid the problem with crossgen.

**Comment by jkotas** [comment]
> emit only the vzeroupper instructions before & after interop call-sites when `COMPlus_ZapDisable` is set

Or when tiered JIT is on? All hot code is going to be redone with tiered JITing.

**Comment by tannergooding** [comment]
Looking at the latest optimization manual, it looks like even more specifically is that the transition penalty only exists when going from 256-bit AVX code to non-VEX 128-bit code:
![image](https://user-images.githubusercontent.com/10487869/47966290-8f64b400-e005-11e8-88dc-899a2c6eb482.png)

If I read that correctly, then we only need to emit `vzeroupper` in methods that execute 256-bit AVX code (so certain HWIntrinsic functions and `Vector<T>` code).
